File: csharp/CsFind/CsFind/FindOptions.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.Json;
using System.Xml.Linq;

using FindOptionsDictionary = System.Collections.Generic.Dictionary<string, System.Collections.Generic.List<System.Collections.Generic.Dictionary<string,string>>>;

namespace CsFind
{
	public class FindOptions
	{
		private readonly string _findOptionsResource;

		private static readonly Dictionary<string, Action<string, FindSettings>> ArgActionDictionary =
			new Dictionary<string, Action<string,FindSettings>>
				{
					{ "in-archiveext", (s, settings) => settings.AddInArchiveExtension(s) },
					{ "in-archivefilepattern", (s, settings) => settings.AddInArchiveFilePattern(s) },
					{ "in-dirpattern", (s, settings) => settings.AddInDirPattern(s) },
					{ "in-ext", (s, settings) => settings.AddInExtension(s) },
					{ "in-filepattern", (s, settings) => settings.AddInFilePattern(s) },
					{ "in-filetype", (s, settings) => settings.AddInFileType(s) },
					{ "out-archiveext", (s, settings) => settings.AddOutArchiveExtension(s) },
					{ "out-archivefilepattern", (s, settings) => settings.AddOutArchiveFilePattern(s) },
					{ "out-dirpattern", (s, settings) => settings.AddOutDirPattern(s) },
					{ "out-ext", (s, settings) => settings.AddOutExtension(s) },
					{ "out-filepattern", (s, settings) => settings.AddOutFilePattern(s) },
					{ "out-filetype", (s, settings) => settings.AddOutFileType(s) },
					{ "settings-file", SettingsFromFile },
				};

		private static readonly Dictionary<string, Action<bool, FindSettings>> BoolFlagActionDictionary =
			new Dictionary<string, Action<bool, FindSettings>>
				{
					{ "archivesonly", (b, settings) => settings.ArchivesOnly = b },
					{ "colorize", (b, settings) => settings.Colorize = b },
					{ "debug", (b, settings) => settings.Debug = b },
					{ "excludearchives", (b, settings) => settings.IncludeArchives = !b },
					{ "excludehidden", (b, settings) => settings.ExcludeHidden = b },
					{ "help", (b, settings) => settings.PrintUsage = b },
					{ "includearchives", (b, settings) => settings.IncludeArchives = b },
					{ "includehidden", (b, settings) => settings.ExcludeHidden = !b },
					{ "listdirs", (b, settings) => settings.ListDirs = b },
					{ "listfiles", (b, settings) => settings.ListFiles = b },
					{ "maxlastmod", (b, settings) => {
							// TODO: convert to datetime
						}
					},
					{ "maxsize", (b, settings) => {
							// TODO: convert to int
						}
					},
					{ "minlastmod", (b, settings) => {
							// TODO: convert to datetime
						}
					},
					{ "minsize", (b, settings) => {
							// TODO: convert to int
						}
					},
					{ "nocolorize", (b, settings) => settings.Colorize = !b },
					{ "norecursive", (b, settings) => settings.Recursive = !b },
					{ "recursive", (b, settings) => settings.Recursive = b },
					{ "verbose", (b, settings) => settings.Verbose = b },
					{ "version", (b, settings) => settings.PrintVersion = b },
				};

		public List<FindOption> Options { get; }
		public Dictionary<string, FindOption> ArgDictionary { get; }
		public Dictionary<string, FindOption> FlagDictionary { get; }

		public FindOptions()
		{
			// _findOptionsResource = EmbeddedResource.GetResourceFileContents("CsFind.Resources.findoptions.xml");
			_findOptionsResource = EmbeddedResource.GetResourceFileContents("CsFind.Resources.findoptions.json");
			Options = new List<FindOption>();
			ArgDictionary = new Dictionary<string, FindOption>();
			FlagDictionary = new Dictionary<string, FindOption>();
			// SetOptionsFromXml();
			SetOptionsFromJson();
		}

		private void SetOptionsFromJson()
		{
			var findOptionsDict = JsonSerializer.Deserialize<FindOptionsDictionary>(_findOptionsResource);
			var optionDicts = findOptionsDict["findoptions"];
			foreach (var optionDict in optionDicts)
			{
				var longArg = optionDict["long"];
				var shortArg = optionDict.ContainsKey("short") ? optionDict["short"] : null;
				var desc = optionDict["desc"];
				if (ArgActionDictionary.ContainsKey(longArg))
				{
					var option = new FindArgOption(shortArg, longArg, ArgActionDictionary[longArg], desc);
					Options.Add(option);
					ArgDictionary.Add(longArg, option);
					if (!string.IsNullOrWhiteSpace(shortArg))
					{
						ArgDictionary.Add(shortArg, option);
					}
				}
				else if (BoolFlagActionDictionary.ContainsKey(longArg))
				{
					var option = new FindFlagOption(shortArg, longArg, BoolFlagActionDictionary[longArg], desc);
					Options.Add(option);
					FlagDictionary.Add(longArg, option);
					if (!string.IsNullOrWhiteSpace(shortArg))
					{
						FlagDictionary.Add(shortArg, option);
					}
				}
			}
		}

		private void SetOptionsFromXml()
		{
			var doc = XDocument.Parse(_findOptionsResource);
			foreach (var f in doc.Descendants("findoption"))
			{
				var longArg = f.Attributes("long").First().Value;
				var shortArg = f.Attributes("short").First().Value;
				var desc = f.Value.Trim();
				if (ArgActionDictionary.ContainsKey(longArg))
				{
					var option = new FindArgOption(shortArg, longArg, ArgActionDictionary[longArg], desc);
					Options.Add(option);
					ArgDictionary.Add(longArg, option);
					if (!string.IsNullOrWhiteSpace(shortArg))
					{
						ArgDictionary.Add(shortArg, option);
					}
				}
				else if (BoolFlagActionDictionary.ContainsKey(longArg))
				{
					var option = new FindFlagOption(shortArg, longArg, BoolFlagActionDictionary[longArg], desc);
					Options.Add(option);
					FlagDictionary.Add(longArg, option);
					if (!string.IsNullOrWhiteSpace(shortArg))
					{
						FlagDictionary.Add(shortArg, option);
					}
				}
			}
		}

		private static void SettingsFromFile(string filePath, FindSettings settings)
		{
			var fileInfo = new FileInfo(filePath);
			if (!fileInfo.Exists)
				throw new FindException("Settings fie not found: " + filePath);
			var contents = FileUtil.GetFileContents(filePath, Encoding.Default);
			SettingsFromJson(contents, settings);
		}

		public static void SettingsFromJson(string jsonString, FindSettings settings)
		{
			var settingsDict = JsonSerializer.Deserialize<Dictionary<string, object>>(jsonString);
			foreach (var (key, value) in settingsDict)
			{
				var obj = (JsonElement)value;
				ApplySetting(key, obj, settings);
			}
		}

		private static void ApplySetting(string arg, JsonElement obj, FindSettings settings)
		{
			switch (obj.ValueKind)
			{
				case JsonValueKind.String:
					ApplySetting(arg, obj.GetString(), settings);
					break;
				case JsonValueKind.True:
					ApplySetting(arg, true, settings);
					break;
				case JsonValueKind.False:
					ApplySetting(arg, false, settings);
					break;
				case JsonValueKind.Number:
					ApplySetting(arg, obj.GetInt32().ToString(), settings);
					break;
				case JsonValueKind.Array:
					foreach (var arrVal in obj.EnumerateArray())
					{
						ApplySetting(arg, arrVal, settings);
					}
					break;
				case JsonValueKind.Undefined:
				case JsonValueKind.Object:
				case JsonValueKind.Null:
				default:
					break;
			}
		}

		private static void ApplySetting(string arg, string val, FindSettings settings)
		{
			if (ArgActionDictionary.ContainsKey(arg))
			{
				ArgActionDictionary[arg](val, settings);
			}
			else if (arg.Equals("path"))
			{
				settings.Paths.Add(val);
			}
			else
			{
				throw new FindException("Invalid option: " + arg);
			}
		}

		private static void ApplySetting(string arg, bool val, FindSettings settings)
		{
			if (BoolFlagActionDictionary.ContainsKey(arg))
			{
				BoolFlagActionDictionary[arg](val, settings);
			}
			else
			{
				throw new FindException("Invalid option: " + arg);
			}
		}

		public FindSettings SettingsFromArgs(IEnumerable<string> args)
		{
			// default to ListFiles = true since this is called from CLI
			var settings = new FindSettings {ListFiles = true};
			var queue = new Queue<string>(args);

			while (queue.Count > 0)
			{
				var s = queue.Dequeue();
				if (s.StartsWith("-"))
				{
					try
					{
						while (s.StartsWith("-"))
						{
							s = s.Substring(1);
						}
					}
					catch (InvalidOperationException e)
					{
						throw new FindException(e.Message);
					}
					if (string.IsNullOrWhiteSpace(s))
					{
						throw new FindException("Invalid option: -");
					}
					if (ArgDictionary.ContainsKey(s))
					{
						try
						{
							((FindArgOption)ArgDictionary[s]).Action(queue.Dequeue(), settings);
						}
						catch (InvalidOperationException e)
						{
							throw new FindException(e.Message);
						}
					}
					else if (FlagDictionary.ContainsKey(s))
					{
						try
						{
							((FindFlagOption)FlagDictionary[s]).Action(true, settings);
						}
						catch (InvalidOperationException e)
						{
							throw new FindException(e.Message);
						}
					}
					else
					{
						throw new FindException("Invalid option: " + s);
					}
				}
				else
				{
					settings.Paths.Add(s);
				}
			}
			return settings;
		}

		public void Usage(int exitCode = 0)
		{
			Console.WriteLine(GetUsageString());
			Environment.Exit(exitCode);
		}

		public string GetUsageString()
		{
			var sb = new StringBuilder();
			sb.AppendLine("\nUsage:");
			sb.AppendLine(" csfind [options] <path> [<path> ...]\n");
			sb.AppendLine("Options:");
			var optStrings = new List<string>();
			var optDescs = new List<string>();
			var longest = 0;
			foreach (var opt in Options.OrderBy(o => o.SortArg))
			{
				var optString = new StringBuilder();
				if (!string.IsNullOrWhiteSpace(opt.ShortArg))
				{
					optString.Append("-" + opt.ShortArg + ",");
				}
				optString.Append("--" + opt.LongArg);
				if (optString.Length > longest)
				{
					longest = optString.Length;
				}
				optStrings.Add(optString.ToString());
				optDescs.Add(opt.Description);
			}
			var format = " {0,-" + longest + "}  {1}";
			for (var i = 0; i < optStrings.Count; i++)
			{
				sb.AppendLine(string.Format(format, optStrings[i], optDescs[i]));
			}
			return sb.ToString();
		}
	}
}

File: java/javafind/src/main/java/javafind/FindOptions.java
/*******************************************************************************
FindOptions

Class to encapsulate all command line find options

@author Cary Clark &lt;clarkcb@gmail.com&gt;
@version $Rev$
@copyright Cary Clark 2012
*******************************************************************************/

package javafind;

import org.json.JSONArray;
import org.json.JSONObject;
import org.json.JSONTokener;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;
import java.util.stream.Collectors;

public class FindOptions {
    private static final String FIND_OPTIONS_JSON_PATH = "/findoptions.json";
    private final List<FindOption> options;
    private final Map<String, String> longArgMap = new HashMap<>();


    public FindOptions() throws IOException {
        options = new ArrayList<>();
        setOptionsFromJson();
    }

    @FunctionalInterface
    private interface BooleanSetter {
        void set(Boolean b, FindSettings settings);
    }

    private final int boolActionMapSize = 18;
    private final Map<String, BooleanSetter> boolActionMap = new HashMap<>(boolActionMapSize) {
        {
            put("archivesonly", (b, settings) -> settings.setArchivesOnly(b));
            put("debug", (b, settings) -> settings.setDebug(b));
            put("excludehidden", (b, settings) -> settings.setIncludeHidden(!b));
            put("followsymlinks", (b, settings) -> settings.setFollowSymlinks(b));
            put("help", (b, settings) -> settings.setPrintUsage(b));
            put("includearchives", (b, settings) -> settings.setIncludeArchives(b));
            put("includehidden", (b, settings) -> settings.setIncludeHidden(b));
            put("nofollowsymlinks", (b, settings) -> settings.setFollowSymlinks(!b));
            put("noprintdirs", (b, settings) -> settings.setPrintDirs(!b));
            put("noprintfiles", (b, settings) -> settings.setPrintFiles(!b));
            put("norecursive", (b, settings) -> settings.setRecursive(!b));
            put("printdirs", (b, settings) -> settings.setPrintDirs(b));
            put("printfiles", (b, settings) -> settings.setPrintFiles(b));
            put("recursive", (b, settings) -> settings.setRecursive(b));
            put("sort-ascending", (b, settings) -> settings.setSortDescending(!b));
            put("sort-caseinsensitive", (b, settings) -> settings.setSortCaseInsensitive(b));
            put("sort-casesensitive", (b, settings) -> settings.setSortCaseInsensitive(!b));
            put("sort-descending", (b, settings) -> settings.setSortDescending(b));
            put("verbose", (b, settings) -> settings.setVerbose(b));
            put("version", (b, settings) -> settings.setPrintVersion(b));
        }
    };

    @FunctionalInterface
    private interface StringSetter {
        void set(String s, FindSettings settings);
    }

    private final int stringActionMapSize = 17;
    private final Map<String, StringSetter> stringActionMap = new HashMap<>(stringActionMapSize) {
        {
            put("in-archiveext", (s, settings) -> settings.addInArchiveExtension(s));
            put("in-archivefilepattern", (s, settings) -> settings.addInArchiveFilePattern(s));
            put("in-dirpattern", (s, settings) -> settings.addInDirPattern(s));
            put("in-ext", (s, settings) -> settings.addInExtension(s));
            put("in-filepattern", (s, settings) -> settings.addInFilePattern(s));
            put("in-filetype", (s, settings) -> settings.addInFileType(s));
            put("maxlastmod", (s, settings) -> settings.setMaxLastMod(s));
            put("minlastmod", (s, settings) -> settings.setMinLastMod(s));
            put("out-archiveext", (s, settings) -> settings.addOutArchiveExtension(s));
            put("out-archivefilepattern", (s, settings) -> settings.addOutArchiveFilePattern(s));
            put("out-dirpattern", (s, settings) -> settings.addOutDirPattern(s));
            put("out-ext", (s, settings) -> settings.addOutExtension(s));
            put("out-filepattern", (s, settings) -> settings.addOutFilePattern(s));
            put("out-filetype", (s, settings) -> settings.addOutFileType(s));
            put("path", (s, settings) -> settings.addPath(s));
            // put("settings-file", (s, settings) -> settingsFromFilePath(s, settings));
            put("sort-by", (s, settings) -> settings.setSortBy(SortBy.forName(s)));
        }
    };

    @FunctionalInterface
    private interface IntegerSetter {
        void set(Integer i, FindSettings settings);
    }

    private final int intActionMapSize = 2;
    private final Map<String, IntegerSetter> intActionMap = new HashMap<>(intActionMapSize) {
        {
            put("maxdepth", (i, settings) -> settings.setMaxDepth(i));
            put("mindepth", (i, settings) -> settings.setMinDepth(i));
        }
    };

    @FunctionalInterface
    private interface LongSetter {
        void set(Long l, FindSettings settings);
    }

    private final int longActionMapSize = 2;
    private final Map<String, LongSetter> longActionMap = new HashMap<>(longActionMapSize) {
        {
            put("maxsize", (l, settings) -> settings.setMaxSize(l));
            put("minsize", (l, settings) -> settings.setMinSize(l));
        }
    };

    private void setOptionsFromJson() {
        var findOptionsInputStream = getClass().getResourceAsStream(FIND_OPTIONS_JSON_PATH);
        assert findOptionsInputStream != null;
        var jsonObj = new JSONObject(new JSONTokener(findOptionsInputStream));
        var findOptionsArray = jsonObj.getJSONArray("findoptions");

        for (int i=0; i<findOptionsArray.length(); i++) {
            var findOptionObj = findOptionsArray.getJSONObject(i);
            var longArg = findOptionObj.getString("long");
            longArgMap.put(longArg, longArg);
            var desc = findOptionObj.getString("desc");
            var shortArg = "";
            if (findOptionObj.has("short")) {
                shortArg = findOptionObj.getString("short");
                longArgMap.put(shortArg, longArg);
            }
            options.add(new FindOption(shortArg, longArg, desc));
        }
    }

    private void applySetting(final String arg, final Object obj, FindSettings settings)
            throws FindException {
        if (this.boolActionMap.containsKey(arg)) {
            if (obj instanceof Boolean) {
                this.boolActionMap.get(arg).set((Boolean)obj, settings);
            } else {
                throw new FindException("Invalid value for option: " + arg);
            }
        } else if (this.stringActionMap.containsKey(arg)) {
            if (obj instanceof String) {
                this.stringActionMap.get(arg).set((String)obj, settings);
            } else if (obj instanceof JSONArray) {
                for (int i=0; i < ((JSONArray)obj).length(); i++) {
                    Object item = ((JSONArray)obj).get(i);
                    if (item instanceof String) {
                        this.stringActionMap.get(arg).set((String)item, settings);
                    } else {
                        throw new FindException("Invalid value for option: " + arg);
                    }
                }
            } else {
                throw new FindException("Invalid value for option: " + arg);
            }
        } else if (this.intActionMap.containsKey(arg)) {
            if (obj instanceof Integer) {
                this.intActionMap.get(arg).set((Integer)obj, settings);
            } else if (obj instanceof Long) {
                this.intActionMap.get(arg).set(((Long)obj).intValue(), settings);
            } else {
                throw new FindException("Invalid value for option: " + arg);
            }
        } else if (this.longActionMap.containsKey(arg)) {
            if (obj instanceof Integer) {
                this.longActionMap.get(arg).set(((Integer)obj).longValue(), settings);
            } else if (obj instanceof Long) {
                this.longActionMap.get(arg).set((Long)obj, settings);
            } else {
                throw new FindException("Invalid value for option: " + arg);
            }
        } else {
            throw new FindException("Invalid option: " + arg);
        }
    }

    public void settingsFromJson(final String json, FindSettings settings) throws FindException {
        var jsonObj = new JSONObject(new JSONTokener(json));
        // keys are sorted so that output is consistent across all versions
        var keys = jsonObj.keySet().stream().sorted().collect(Collectors.toList());
        for (var k : keys) {
            var v = jsonObj.get(k);
            if (v != null) {
                applySetting(k, v, settings);
            }
        }
    }

    private void settingsFromFilePath(final String filePath, FindSettings settings) throws FindException {
        var path = FileUtil.expandPath(Paths.get(filePath));
        if (!Files.exists(path)) {
            throw new FindException("Settings file not found: " + filePath);
        }
        if (!filePath.endsWith(".json")) {
            throw new FindException("Invalid settings file (must be JSON): " + filePath);
        }
        try {
            settingsFromJson(FileUtil.getFileContents(path), settings);
        } catch (FileNotFoundException e) {
            throw new FindException("Settings file not found: " + filePath);
        } catch (IOException e) {
            throw new FindException("IOException reading settings file: " + filePath);
        }
    }

    public final FindSettings settingsFromArgs(final String[] args) throws FindException {
        var settings = new FindSettings();
        // default printFiles to true since running from command line
        settings.setPrintFiles(true);

        Queue<String> queue = new LinkedList<>(Arrays.asList(args));
        while (!queue.isEmpty()) {
            var arg = queue.remove();
            if (arg.startsWith("-")) {
                while (arg.startsWith("-")) {
                    arg = arg.substring(1);
                }
                var longArg = longArgMap.get(arg);
                if (this.boolActionMap.containsKey(longArg)) {
                    this.boolActionMap.get(longArg).set(true, settings);
                } else if (this.stringActionMap.containsKey(longArg)
                        || this.intActionMap.containsKey(longArg)
                        || this.longActionMap.containsKey(longArg)
                        || longArg.equals("settings-file")) {
                    if (!queue.isEmpty()) {
                        String argVal = queue.remove();
                        if (this.stringActionMap.containsKey(longArg)) {
                            this.stringActionMap.get(longArg).set(argVal, settings);
                        } else if (this.intActionMap.containsKey(longArg)) {
                            this.intActionMap.get(longArg).set(Integer.parseInt(argVal), settings);
                        } else if (this.longActionMap.containsKey(longArg)) {
                            this.longActionMap.get(longArg).set(Long.parseLong(argVal), settings);
                        } else if (longArg.equals("settings-file")) {
                            settingsFromFilePath(argVal, settings);
                        }
                    } else {
                        throw new FindException("Missing value for option " + arg);
                    }
                } else {
                    throw new FindException("Invalid option: " + arg);
                }
            } else {
                settings.addPath(arg);
            }
        }
        return settings;
    }

    public final void usage(final int exitStatus) {
        System.out.println(this.getUsageString());
        System.exit(exitStatus);
    }

    public final String getUsageString() {
        var sb = new StringBuilder();
        sb.append("Usage:\n");
        sb.append(" javafind [options] <path> [<path> ...]\n\n");
        sb.append("Options:\n");

        this.options.sort(Comparator.comparing(FindOption::getSortArg));

        var optStrings = new ArrayList<>();
        var optDescs = new ArrayList<>();
        int longest = 0;
        for (var opt : this.options) {
            var optString = new StringBuilder();
            var shortArg = opt.getShortArg();
            if (null != shortArg && !shortArg.isEmpty()) {
                optString.append("-").append(shortArg).append(",");
            }
            optString.append("--").append(opt.getLongArg());
            if (optString.length() > longest) {
                longest = optString.length();
            }
            optStrings.add(optString.toString());
            optDescs.add(opt.getDescription());
        }
        final var format = " %1$-" + longest + "s  %2$s\n";
        for (int i = 0; i < optStrings.size(); i++) {
            sb.append(String.format(format, optStrings.get(i), optDescs.get(i)));
        }
        return sb.toString();
    }
}
